### ISSUE TITLE
Bleadperl v5.17.3-44-g7261499 breaks ZEFRAM/Scope-Escape-0.004.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12344]
Migrated from [rt.perl.org#114556](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114556) (status was 'resolved')

Searchable as RT114556$

**Comment by p5pRT** [comment]
### From @andk

Git bisect


---
7261499db89d7afd6c64079406dc32f10acfe512 is the first bad commit
commit 7261499db89d7afd6c64079406dc32f10acfe512
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Thu Aug 16 16&#8203;:46&#8203;:20 2012 \-0700

&nbsp;    Stop padlists from being AVs

also affected


---
VPIT/Scope\-Upper\-0\.18
ZEFRAM/Sub\-Mutate\-0\.004

perl \-V


---
Summary of my perl5 \(revision 5 version 17 subversion 4\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 7261499db89d7afd6c64079406dc32f10acfe512
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.2\.0\-3\-amd64\, archname=x86\_64\-linux
&nbsp;    uname='linux k83 3\.2\.0\-3\-amd64 \#1 smp mon jul 23 02&#8203;:45&#8203;:17 utc 2012 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/src/perl/repoperls/installed\-perls/perl/v5\.17\.3\-44\-g7261499/165a \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Uuseithreads \-Uuselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=undef\, usemultiplicity=undef
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.7\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.13'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES PERLIO\_LAYERS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_MALLOC\_WRAP PERL\_PRESERVE\_IVUV PERL\_USE\_DEVEL
&nbsp;                        USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;                        USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_PERLIO USE\_PERL\_ATOF
&nbsp;  Built under linux
&nbsp;  Compiled at Aug 23 2012 08&#8203;:32&#8203;:36
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.3\-44\-g7261499/165a/lib/site\_perl/5\.17\.4/x86\_64\-linux
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.3\-44\-g7261499/165a/lib/site\_perl/5\.17\.4
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.3\-44\-g7261499/165a/lib/5\.17\.4/x86\_64\-linux
&nbsp;    /home/src/perl/repoperls/installed\-perls/perl/v5\.17\.3\-44\-g7261499/165a/lib/5\.17\.4
&nbsp;    \.



\-\- 
andreas

**Comment by p5pRT** [comment]
### From perl@profvince.com

On 23/08/2012 21&#8203;:20\, \(Andreas J\. Koenig\) \(via RT\) wrote&#8203;:
> \# New Ticket Created by  \(Andreas J\. Koenig\)
> \# Please include the string&#8203;:  \[perl \#114556\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114556 >
>
>
> Git bisect
> \-\-\-\-\-\-\-\-\-\-
> 7261499db89d7afd6c64079406dc32f10acfe512 is the first bad commit
> commit 7261499db89d7afd6c64079406dc32f10acfe512
> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> Date&#8203;:   Thu Aug 16 16&#8203;:46&#8203;:20 2012 \-0700
>
>      Stop padlists from being AVs
>
> also affected
> \-\-\-\-\-\-\-\-\-\-\-\-\-
> VPIT/Scope\-Upper\-0\.18


Thanks for the report\. It has already been fixed in git\.


Vincent

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Aug 23 12&#8203;:20&#8203;:33 2012\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> Git bisect
> \-\-\-\-\-\-\-\-\-\-
> 7261499db89d7afd6c64079406dc32f10acfe512 is the first bad commit
> commit 7261499db89d7afd6c64079406dc32f10acfe512
> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> Date&#8203;:   Thu Aug 16 16&#8203;:46&#8203;:20 2012 \-0700
> 
>     Stop padlists from being AVs
> 
> also affected
> \-\-\-\-\-\-\-\-\-\-\-\-\-
> VPIT/Scope\-Upper\-0\.18
> ZEFRAM/Sub\-Mutate\-0\.004

Here is a more complete list&#8203;:

Devel&#8203;::FindRef \-\- passes its tests\, but still probably broken

Broken; unpatched&#8203;:

B&#8203;::C \- broken already due to \->op\_attached
Coro \- tricky
Lexical&#8203;::Var \- tricky
Scope&#8203;::Escape \- tricky
Sub&#8203;::Filter \- depensd on Sub&#8203;::Mutate; maybe needs its own patching too
Sub&#8203;::Mutate \- tricky

Broken; patched or patch submitted&#8203;:

Data&#8203;::Dump&#8203;::Streamer \- unrelated to PADLIST actually; see \#79127
Devel&#8203;::LexAlias \- \#79267
Eval&#8203;::Compile \- \#79153
fields&#8203;::aliased \- \#79276
Parse&#8203;::Perl \- \#79158
Scope&#8203;::Upper \- by its author
Sub&#8203;::Clone \- \#79266

Fixed&#8203;:

B&#8203;::Generate \- \#79155
PadWalker \- \#79154


‘Tricky’ modules are those that store pad stuff in AVs or create their
own padlists/names\.  So it’s not just a matter of substitution new
macros for old\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 31 14&#8203;:07&#8203;:10 2012\, sprout wrote&#8203;:
> ‘Tricky’ modules are those that store pad stuff in AVs or create their
> own padlists/names\.  So it’s not just a matter of substitution new
> macros for old\.

I’m always making that sort of mistake\.  I meant ‘substituting’ of
course\.  My brain is saying one thing\, but my fingers often have their
own ideas about inflections\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 31 14&#8203;:07&#8203;:10 2012\, sprout wrote&#8203;:
> Devel&#8203;::FindRef \-\- passes its tests\, but still probably broken

Indeed\, it was broken by 7261499\.  Here is a patch for it\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

<details><summary>Inline Patch</summary>

```diff
diff -Nurp Devel-FindRef-1.422-qp6V8T/FindRef.xs Devel-FindRef-1.422-um5Hdd/FindRef.xs
--- Devel-FindRef-1.422-qp6V8T/FindRef.xs	2009-08-30 07:33:01.000000000 -0700
+++ Devel-FindRef-1.422-um5Hdd/FindRef.xs	2012-10-17 16:16:12.000000000 -0700
@@ -21,6 +21,20 @@
 # define GvNAME_HEK(sv) 1
 #endif
 
+#ifndef PadARRAY
+typedef SV PADNAME;
+# if PERL_VERSION < 8 || (PERL_VERSION == 8 && !PERL_SUBVERSION)
+typedef AV PADLIST;
+typedef AV PAD;
+# endif
+# define PadlistARRAY(pl)	((PAD **)AvARRAY(pl))
+# define PadlistMAX(pl)		AvFILLp(pl)
+# define PadlistNAMESARRAY(pl)	AvARRAY(*PadlistARRAY(pl))
+# define PadARRAY		AvARRAY
+# define PadMAX			AvFILLp
+# define PadnamePV(pn)		(SvPOKp(pn) ? SvPVX(pn) : NULL)
+#endif
+
 #define res_pair(text)						\
   do {								\
     AV *av = newAV ();						\
@@ -160,26 +174,27 @@ find_ (SV *target_ref)
 
                               if (depth)
                                 {
-                                  AV *padlist = CvPADLIST (sv);
+                                  PADLIST *padlist = CvPADLIST (sv);
 
                                   while (depth)
                                     {
-                                      AV *pad = (AV *)AvARRAY (padlist)[depth];
+                                      PAD *pad = PadlistARRAY (padlist)[depth];
 
                                       av_push (excl, newSVuv (PTR2UV (pad))); /* exclude pads themselves from being found */
 
                                       /* The 0th pad slot is @_ */
-                                      if (AvARRAY (pad)[0] == targ)
+                                      if (PadARRAY (pad)[0] == targ)
                                         res_pair ("the argument array for");
 
-                                      for (i = AvFILLp (pad) + 1; --i; )
-                                        if (AvARRAY (pad)[i] == targ)
+                                      for (i = PadMAX (pad) + 1; --i; )
+                                        if (PadARRAY (pad)[i] == targ)
                                           {
                                             /* Values from constant functions are stored in the pad without any name */
-                                            SV *name_sv = AvARRAY (AvARRAY (padlist)[0])[i];
+                                            PADNAME *name =
+                                             PadlistNAMESARRAY(padlist)[i];
 
-                                            if (name_sv && SvPOK (name_sv))
-                                              res_pair (form ("the lexical '%s' in", SvPVX (name_sv)));
+                                            if (name && PadnamePV (name))
+                                              res_pair (form ("the lexical '%s' in", PadnamePV(name)));
                                             else
                                               res_pair ("an unnamed lexical in");
                                           }
diff -Nurp Devel-FindRef-1.422-qp6V8T/MANIFEST Devel-FindRef-1.422-um5Hdd/MANIFEST
--- Devel-FindRef-1.422-qp6V8T/MANIFEST	2009-08-30 07:34:31.000000000 -0700
+++ Devel-FindRef-1.422-um5Hdd/MANIFEST	2012-10-17 14:08:52.000000000 -0700
@@ -6,4 +6,5 @@ README
 FindRef.xs
 FindRef.pm
 t/00_load.t
+t/pad.t
 META.yml                                 Module meta-data (added by MakeMaker)
diff -Nurp Devel-FindRef-1.422-qp6V8T/t/pad.t Devel-FindRef-1.422-um5Hdd/t/pad.t
--- Devel-FindRef-1.422-qp6V8T/t/pad.t	1969-12-31 16:00:00.000000000 -0800
+++ Devel-FindRef-1.422-um5Hdd/t/pad.t	2012-10-17 14:18:14.000000000 -0700
@@ -0,0 +1,10 @@
+BEGIN { $| = 1; print "1..1\n"; }
+use Devel::FindRef;
+my $x = [];
+print "not " unless Devel::FindRef::track($x) =~
+/^ARRAY\(0x[\da-f]+\) \[refcount \d+] is
+referenced by REF\(0x[\da-f]+\) \[refcount \d+], which is
+   the lexical '\$x' in CODE\(0x[\da-f]+\) \[refcount \d+], which is
+      the main body of the program\.
+\z/;
+print "ok 1\n";
```
</details>

**Comment by p5pRT** [comment]
### From @rurban

On Fri\, Aug 31\, 2012 at 4&#8203;:07 PM\, Father Chrysostomos via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Thu Aug 23 12&#8203;:20&#8203;:33 2012\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
>> Git bisect
>> \-\-\-\-\-\-\-\-\-\-
>> 7261499db89d7afd6c64079406dc32f10acfe512 is the first bad commit
>> commit 7261499db89d7afd6c64079406dc32f10acfe512
>> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
>> Date&#8203;:   Thu Aug 16 16&#8203;:46&#8203;:20 2012 \-0700
>>
>>     Stop padlists from being AVs
>>
>> also affected
>> \-\-\-\-\-\-\-\-\-\-\-\-\-
>> VPIT/Scope\-Upper\-0\.18
>> ZEFRAM/Sub\-Mutate\-0\.004
>
> Here is a more complete list&#8203;:
>
> Devel&#8203;::FindRef \-\- passes its tests\, but still probably broken
>
> Broken; unpatched&#8203;:
>
> B&#8203;::C \- broken already due to \->op\_attached

Fixed lately for this and the new PADLIST type\, but still broken for
B&#8203;::Bytecode\, B&#8203;::C and B&#8203;::CC\.

I am still wondering if I can keep the static assignment of padlists
as per pre\-5\.17\.5 or if
I have to ressort back to a run\-time Perl\_pad\_new\(0\) call\. The recent
ptr as unique value change for
pad\_generation needs one new run\-time call per pad\, but this is not a
dramatic slowdown\.
I like that small pads do not cost at least 3 AV entries anymore\, but
with B&#8203;::C this was no issue\,
since my array entries were static\.

> Coro \- tricky
> Lexical&#8203;::Var \- tricky
> Scope&#8203;::Escape \- tricky
> Sub&#8203;::Filter \- depensd on Sub&#8203;::Mutate; maybe needs its own patching too
> Sub&#8203;::Mutate \- tricky
>
> Broken; patched or patch submitted&#8203;:
>
> Data&#8203;::Dump&#8203;::Streamer \- unrelated to PADLIST actually; see \#79127
> Devel&#8203;::LexAlias \- \#79267
> Eval&#8203;::Compile \- \#79153
> fields&#8203;::aliased \- \#79276
> Parse&#8203;::Perl \- \#79158
> Scope&#8203;::Upper \- by its author
> Sub&#8203;::Clone \- \#79266
>
> Fixed&#8203;:
>
> B&#8203;::Generate \- \#79155
> PadWalker \- \#79154
>
> ‘Tricky’ modules are those that store pad stuff in AVs or create their
> own padlists/names\.  So it’s not just a matter of substitution new
> macros for old\.
\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From schmorp@schmorp.de

On Wed\, Oct 17\, 2012 at 04&#8203;:23&#8203;:54PM \-0700\, Father Chrysostomos via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Fri Aug 31 14&#8203;:07&#8203;:10 2012\, sprout wrote&#8203;:
> > Devel&#8203;::FindRef \-\- passes its tests\, but still probably broken
> 
> Indeed\, it was broken by 7261499\.  Here is a patch for it\.

Thanks\, I'll look at it when I make the next release \(but it looks
obviously fine\) :\)

\-\- 
&nbsp;                The choice of a       Deliantra\, the free code\+content MORPG
&nbsp;      \-\-\-\-\-==\-     \_GNU\_              http&#8203;://www\.deliantra\.net
&nbsp;      \-\-\-\-==\-\- \_       generation
&nbsp;      \-\-\-==\-\-\-\(\_\)\_\_  \_\_ \_\_\_\_  \_\_      Marc Lehmann
&nbsp;      \-\-==\-\-\-/ / \_ \\/ // /\\ \\/ /      schmorp@&#8203;schmorp\.de
&nbsp;      \-=====/\_/\_//\_/\\\_\,\_/ /\_/\\\_\\

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 31 14&#8203;:07&#8203;:10 2012\, sprout wrote&#8203;:
> Broken; unpatched&#8203;:
\.\.\.
> Coro \- tricky

Here is a patch for Coro\.  It uses a mallocked array to hold a list of
padlists in magic\, instead of an AV\, since padlists are no longer SV\-ish\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

<details><summary>Inline Patch</summary>

```diff
diff -rup Coro-6.10-bbke1u/Coro/State.xs Coro-6.10-ibjGm2/Coro/State.xs
--- Coro-6.10-bbke1u/Coro/State.xs	2012-10-08 16:10:49.000000000 -0700
+++ Coro-6.10-ibjGm2/Coro/State.xs	2012-10-22 23:01:05.000000000 -0700
@@ -27,6 +27,24 @@
 # define SVs_PADSTALE 0
 #endif
 
+#ifdef PadARRAY
+# define NEWPADAPI
+# define newPADLIST(var)	(Newz(1234,var,1,PADLIST), \
+				 Newx(PadlistARRAY(var), 2, PAD *))
+#else
+typedef AV PADNAMELIST;
+# if PERL_VERSION < 8 || (PERL_VERSION == 8 && !PERL_SUBVERSION)
+typedef AV PADLIST;
+typedef AV PAD;
+# endif
+# define PadlistARRAY(pl)	((PAD **)AvARRAY(pl))
+# define PadlistMAX(pl)		AvFILLp(pl)
+# define PadlistNAMES(pl)	(*PadlistARRAY(pl))
+# define PadARRAY		AvARRAY
+# define PadMAX			AvFILLp
+# define newPADLIST(var)	((var) = newAV(), av_extend(var, 1))
+#endif
+
 #if defined(_WIN32)
 # undef HAS_GETTIMEOFDAY
 # undef setjmp
@@ -526,71 +544,80 @@ SvSTATE_ (pTHX_ SV *coro)
 /*****************************************************************************/
 /* padlist management and caching */
 
-ecb_inline AV *
+ecb_inline PADLIST *
 coro_derive_padlist (pTHX_ CV *cv)
 {
-  AV *padlist = CvPADLIST (cv);
-  AV *newpadlist, *newpad;
-
-  newpadlist = newAV ();
+  PADLIST *padlist = CvPADLIST (cv);
+  PADLIST *newpadlist;
+  PAD *newpad;
+  PADOFFSET const off = PadlistMAX (padlist) + 1;
+
+  newPADLIST(newpadlist);
+#if !PERL_VERSION_ATLEAST(5,15,3)
+  /* Padlists are AvREAL as of 5.15.3. See perl bug #98092 and perl
+     commit 7d953ba. */
   AvREAL_off (newpadlist);
+#endif
 #if PERL_VERSION_ATLEAST (5,10,0)
-  Perl_pad_push (aTHX_ padlist, AvFILLp (padlist) + 1);
+  Perl_pad_push (aTHX_ padlist, off);
 #else
-  Perl_pad_push (aTHX_ padlist, AvFILLp (padlist) + 1, 1);
+  Perl_pad_push (aTHX_ padlist, off, 1);
 #endif
-  newpad = (AV *)AvARRAY (padlist)[AvFILLp (padlist)];
-  --AvFILLp (padlist);
+  newpad = PadlistARRAY (padlist)[off];
+  PadlistMAX (padlist) = off-1;
 
-  av_store (newpadlist, 0, SvREFCNT_inc_NN (AvARRAY (padlist)[0]));
-  av_store (newpadlist, 1, (SV *)newpad);
+  /* Already extended to 2 elements by newPADLIST. */
+  PadlistMAX (newpadlist) = 1;
+  PadlistNAMES (newpadlist) =
+    (PADNAMELIST *)SvREFCNT_inc_NN (PadlistNAMES (padlist));
+  PadlistARRAY (newpadlist)[1] = newpad;
 
   return newpadlist;
 }
 
 ecb_inline void
-free_padlist (pTHX_ AV *padlist)
+free_padlist (pTHX_ PADLIST *padlist)
 {
   /* may be during global destruction */
   if (!IN_DESTRUCT)
     {
-      I32 i = AvFILLp (padlist);
+      I32 i = PadlistMAX (padlist);
 
       while (i > 0) /* special-case index 0 */
         {
           /* we try to be extra-careful here */
-          AV *av = (AV *)AvARRAY (padlist)[i--];
-          I32 j = AvFILLp (av);
+          PAD *pad = PadlistARRAY (padlist)[i--];
+          I32 j = PadMAX (pad);
 
           while (j >= 0)
-            SvREFCNT_dec (AvARRAY (av)[j--]);
+            SvREFCNT_dec (PadARRAY (pad)[j--]);
 
-          AvFILLp (av) = -1;
-          SvREFCNT_dec (av);
+          PadMAX (pad) = -1;
+          SvREFCNT_dec (pad);
         }
 
-      SvREFCNT_dec (AvARRAY (padlist)[0]);
+      SvREFCNT_dec (PadlistNAMES (padlist));
 
+#ifdef NEWPADAPI
+      Safefree(PadlistARRAY(padlist));
+      Safefree(padlist);
+#else
       AvFILLp (padlist) = -1;
+      AvREAL_off(padlist);
       SvREFCNT_dec ((SV*)padlist);
+#endif
     }
 }
 
 static int
 coro_cv_free (pTHX_ SV *sv, MAGIC *mg)
 {
-  AV *padlist;
-  AV *av = (AV *)mg->mg_obj;
-
-  /* perl manages to free our internal AV and _then_ call us */
-  if (IN_DESTRUCT)
-    return 0;
-
-  /* casting is fun. */
-  while (&PL_sv_undef != (SV *)(padlist = (AV *)av_pop (av)))
-    free_padlist (aTHX_ padlist);
+  PADLIST *padlist;
+  PADLIST **padlists = (PADLIST **)(mg->mg_ptr + sizeof(size_t));
+  size_t len = *(size_t *)mg->mg_ptr;
 
-  SvREFCNT_dec (av); /* sv_magicext increased the refcount */
+  while (len--)
+    free_padlist (aTHX_ padlists[len]);
 
   return 0;
 }
@@ -605,10 +632,10 @@ ecb_inline void
 get_padlist (pTHX_ CV *cv)
 {
   MAGIC *mg = CORO_MAGIC_cv (cv);
-  AV *av;
+  size_t *lenp;
 
-  if (ecb_expect_true (mg && AvFILLp ((av = (AV *)mg->mg_obj)) >= 0))
-    CvPADLIST (cv) = (AV *)AvARRAY (av)[AvFILLp (av)--];
+  if (ecb_expect_true (mg && *(lenp = (size_t *)mg->mg_ptr)))
+    CvPADLIST (cv) = ((PADLIST **)(mg->mg_ptr + sizeof(size_t)))[--*lenp];
   else
    {
 #if CORO_PREFER_PERL_FUNCTIONS
@@ -628,17 +655,20 @@ ecb_inline void
 put_padlist (pTHX_ CV *cv)
 {
   MAGIC *mg = CORO_MAGIC_cv (cv);
-  AV *av;
 
   if (ecb_expect_false (!mg))
-    mg = sv_magicext ((SV *)cv, (SV *)newAV (), CORO_MAGIC_type_cv, &coro_cv_vtbl, 0, 0);
-
-  av = (AV *)mg->mg_obj;
-
-  if (ecb_expect_false (AvFILLp (av) >= AvMAX (av)))
-    av_extend (av, AvFILLp (av) + 1);
+  {
+    mg = sv_magicext ((SV *)cv, NULL, CORO_MAGIC_type_cv, &coro_cv_vtbl, 0, 0);
+    Newz(1234,mg->mg_ptr,sizeof(size_t) + sizeof(PADLIST *),char);
+    mg->mg_len = 1; /* so mg_free frees mg_ptr */
+  }
+  else
+    Renew(mg->mg_ptr,
+          sizeof(size_t) + (*(size_t *)mg->mg_ptr + 1) * sizeof(PADLIST *),
+          char);
 
-  AvARRAY (av)[++AvFILLp (av)] = (SV *)CvPADLIST (cv);
+  ((PADLIST **)(mg->mg_ptr + sizeof(size_t)))[(*(size_t *)mg->mg_ptr)++] =
+    CvPADLIST(cv);
 }
 
 /** load & save, init *******************************************************/
@@ -723,7 +753,7 @@ load_perl (pTHX_ Coro__State c)
       {
         put_padlist (aTHX_ cv); /* mark this padlist as available */
         CvDEPTH (cv) = PTR2IV (POPs);
-        CvPADLIST (cv) = (AV *)POPs;
+        CvPADLIST (cv) = (PADLIST *)POPs;
       }
 
     PUTBACK;
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Oct 17 17&#8203;:04&#8203;:00 2012\, rurban wrote&#8203;:
> On Fri\, Aug 31\, 2012 at 4&#8203;:07 PM\, Father Chrysostomos via RT
> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > On Thu Aug 23 12&#8203;:20&#8203;:33 2012\,
> andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> >> Git bisect
> >> \-\-\-\-\-\-\-\-\-\-
> >> 7261499db89d7afd6c64079406dc32f10acfe512 is the first bad commit
> >> commit 7261499db89d7afd6c64079406dc32f10acfe512
> >> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> >> Date&#8203;:   Thu Aug 16 16&#8203;:46&#8203;:20 2012 \-0700
> >>
> >>     Stop padlists from being AVs
> >>
> >> also affected
> >> \-\-\-\-\-\-\-\-\-\-\-\-\-
> >> VPIT/Scope\-Upper\-0\.18
> >> ZEFRAM/Sub\-Mutate\-0\.004
> >
> > Here is a more complete list&#8203;:
> >
> > Devel&#8203;::FindRef \-\- passes its tests\, but still probably broken
> >
> > Broken; unpatched&#8203;:
> >
> > B&#8203;::C \- broken already due to \->op\_attached
> 
> Fixed lately for this and the new PADLIST type\, but still broken for
> B&#8203;::Bytecode\, B&#8203;::C and B&#8203;::CC\.

I had a look at B&#8203;::C myself with the aim of writing a patch\.  But there
is so much to learn there\-\-pages and pages of gibberish\, like perl’s
regexp engine\. :\-\)  Can I leave it to you?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @rurban


On Oct 23\, 2012\, at 11&#8203;:34 PM\, "Father Chrysostomos via RT" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Wed Oct 17 17&#8203;:04&#8203;:00 2012\, rurban wrote&#8203;:
>> On Fri\, Aug 31\, 2012 at 4&#8203;:07 PM\, Father Chrysostomos via RT
>> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>> On Thu Aug 23 12&#8203;:20&#8203;:33 2012\,
>> andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
>>>> Git bisect
>>>> \-\-\-\-\-\-\-\-\-\-
>>>> 7261499db89d7afd6c64079406dc32f10acfe512 is the first bad commit
>>>> commit 7261499db89d7afd6c64079406dc32f10acfe512
>>>> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
>>>> Date&#8203;:   Thu Aug 16 16&#8203;:46&#8203;:20 2012 \-0700
>>>> 
>>>>    Stop padlists from being AVs
>>>> 
>>>> also affected
>>>> \-\-\-\-\-\-\-\-\-\-\-\-\-
>>>> VPIT/Scope\-Upper\-0\.18
>>>> ZEFRAM/Sub\-Mutate\-0\.004
>>> 
>>> Here is a more complete list&#8203;:
>>> 
>>> Devel&#8203;::FindRef \-\- passes its tests\, but still probably broken
>>> 
>>> Broken; unpatched&#8203;:
>>> 
>>> B&#8203;::C \- broken already due to \->op\_attached
>> 
>> Fixed lately for this and the new PADLIST type\, but still broken for
>> B&#8203;::Bytecode\, B&#8203;::C and B&#8203;::CC\.
> 
> I had a look at B&#8203;::C myself with the aim of writing a patch\.  But there
> is so much to learn there\-\-pages and pages of gibberish\, like perl’s
> regexp engine\. :\-\)  Can I leave it to you?

Sure\. I think I am close and I don't need anything else\.

B&#8203;::C is also broken regarding the removal of op\_latefree\(d\)\, which can be circumvented by \-fno\-destruct \(\-O3\)\.
op\_static as in my patch in RT \#115452 needs to be added \(again as with 5\.9\)\.

**Comment by p5pRT** [comment]
### From @cpansprout

I have provided patches for all these modules&#8203;:

Coro
Data&#8203;::Dump&#8203;::Streamer \- unrelated to PADLIST actually; see \#79127
Devel&#8203;::FindRef
Devel&#8203;::LexAlias \- \#79267
fields&#8203;::aliased \- \#79276
Lexical&#8203;::Var \- \#80309
Parse&#8203;::Perl \- \#79158
Scope&#8203;::Escape \- \#80348
Sub&#8203;::Clone \- \#79266
Sub&#8203;::Filter \- \#80367
Sub&#8203;::Mutate \- \#80366

These four have been fixed already&#8203;:

B&#8203;::Generate \- \#79155
Eval&#8203;::Compile \- \#79153
PadWalker \- \#79154
Scope&#8203;::Upper

The unfixed modules with 5 or more dependents are now listed in
perl5180delta\, as of commit 0f023b5a2\.

So that means I can finally close this ticket\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
